### ISSUE TITLE
ZF3 compatibility

### ISSUE THREAD

**Issue by pawel-zawadzki** [neilime/zf2-twb-bundle#176]
Hi, are you planning to release twb-bundle compatibile with ZF3? Right now you're not able to install twb-bundle with ZF3 project because of the zend-mvc version upgrade to 3.*. Composer error:

`- neilime/zf2-twb-bundle 2.6.1 requires zendframework/zend-mvc ~2.3 -> satisfiable by zendframework/zend-mvc[2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.3.5, 2.3.6, 2.3.7, 2.3.8, 2.3.9, 2.4.0, 2.4.0rc1, 2.4.0rc2, 2.4.0rc3, 2.4.0rc4, 2.4.0rc5, 2.4.0rc6, 2.4.0rc7, 2.4.1, 2.4.10, 2.4.2, 2.4.3, 2.4.4, 2.4.5, 2.4.6, 2.4.7, 2.4.8, 2.4.9, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.6.0, 2.6.1, 2.6.2, 2.6.3, 2.7.0, 2.7.1, 2.7.10, 2.7.2, 2.7.3, 2.7.4, 2.7.5, 2.7.6, 2.7.7, 2.7.8, 2.7.9] but these conflict with your requirements or minimum-stability.`

There are probably more compatibility issues with the newest ZF version, but this is the first one.

**Comment by neilime** [comment]
Hi,

This upgrade is planned.

**Comment by necromant2005** [comment]
Temporary migrated to zf3 support, BUT test are not passing jet
fork: https://github.com/necromant2005/zf2-twb-bundle
```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/necromant2005/zf2-twb-bundle"
        }
    ],
       "require": {
            "neilime/zf2-twb-bundle": "dev-master"
    },
```

**Comment by craig-winaust** [comment]
necromant2005 your fork is mostly working for me, except the class "form-control" is not being added to input elements.  Haven't been able to work out why just yet. Any ideas?

**Comment by necromant2005** [comment]
@craig-winaust could you show a sample of your form ? I'll check

**Comment by craig-winaust** [comment]
It's just a standard zf  form with elements like:

```
        $this->add(array(
            'name' => 'credit_card_number',
            'attributes' => array(
                'id'  => 'credit_card_number',
                'type'  => 'text',
                'class' => 'form-control', //had to add the class which used to be automatically added by two
                'maxlength' => 19, 
                'size'  => 19,
            ),
            'options' => array(
                'label' => 'Credit Card Number',
                'column-size' => 'sm-10',
                'label_attributes' => array('class' => 'col-sm-2'),
            ),
        ));
```

Twb would add the class form-control automatically in the zf2 version to form elements. But I had to add the class in the form in the zf3 version. It's a work around I can live with.

**Comment by necromant2005** [comment]
Fixed all tests. But I haven't found why it misses class yet.

**Comment by neilime** [comment]
Thank you @necromant2005 for your work, I fixed the broken phpunit tests in the PR #180. I don't understand the bug with the missing "form-control" class. @craig-winaust can you test with this new branch.

**Comment by pobidowski** [comment]
The same, missing form-control class:

`$element = new Element\Text('name');
$element->setLabel('Nazwa podatku');
$element->setLabelAttributes([
     'class' => 'control-label'
]);
$element->setAttributes([
            'data-ng-model' => 'inputData.name',
            'required' => 'true',
            'placeholder' => 'Nazwa podatku'
        ]);
        $this->add($element);
`

Form is rendering in view:
`echo $this->form($form);`

**Comment by pobidowski** [comment]
The problem is that the `$this->getElementHelper()` in TwbBundleFormRow class returns an instance of `Zend\Form\View\Helper\FormElement` class not an instance of `TwbBundleFormElement` class

**Comment by pobidowski** [comment]
Helped something like this in module config:

`
'view_helpers' => [
        'factories' => [
            \TwbBundle\Form\View\Helper\TwbBundleFormElement::class => \TwbBundle\Form\View\Helper\Factory\TwbBundleFormElementFactory::class
        ],
        'aliases' => [
            'form_element' => \TwbBundle\Form\View\Helper\TwbBundleFormElement::class
        ]
    ]
`

**Comment by necromant2005** [comment]
@pobidowski I've applied your patch. Could you check again necromant2005/zf2-twb-bundle ?

**Comment by pobidowski** [comment]
Thanks, now its working

**Comment by necromant2005** [comment]
great! thank you @pobidowski

**Comment by KiwiJuicer** [comment]
@neilime When do you reckon to support latest ZF2/3 dependencies?

**Comment by exptom** [comment]
@KiwiJuicer the latest master supports ZF3.
@neilime  is there any chance you could tag master?

**Comment by neilime** [comment]
Done : v3.0.0

**Comment by exptom** [comment]
:+1:  thank you @neilime much appreciated.

**Comment by nynka** [comment]
Thanks for tagging @neilime. Is the "missing form-control" issue fixed in v3.0.0?

**Comment by neilime** [comment]
@nynka : Yes it should fix it.
